internal/utils/utils.go

```diff
@@ -52,3 +62,102 @@ func GetOwnerRefFromObj(owner ctrlclient.Object, scheme *runtime.Scheme) (metav1
 	ownerRef.Kind = kind
 	return ownerRef, nil
 }
+
+// LoadCRDsFromDir reads dir to find any CustomResourceDefinition YAML manifest
+// files. It only supports apiextensionsv1.CustomResourceDefinition. Returns a
+// slice of apiextensionsv1.CustomResourceDefinition objects or an error if any
+// occurred.
+func LoadCRDsFromDir(dir string) ([]apiextensionsv1.CustomResourceDefinition, error) {
+	d, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	var crds []apiextensionsv1.CustomResourceDefinition
+	for _, f := range d {
+		if f.IsDir() {
+			continue
+		}
+
+		if !strings.HasSuffix(f.Name(), ".yaml") {
+			continue
+		}
+
+		fn := filepath.Join(dir, f.Name())
+		fh, err := os.Open(fn)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := io.ReadAll(fh)
+		if err != nil {
+			return nil, err
+		}
+
+		jsonB, err := yaml.YAMLToJSONStrict(b)
+		if err != nil {
+			return nil, err
+		}
+
+		var crd apiextensionsv1.CustomResourceDefinition
+		if err := json.Unmarshal(jsonB, &crd); err != nil {
+			return nil, err
+		}
+
+		if crd.GroupVersionKind() != apiextensionsv1.SchemeGroupVersion.WithKind(crd.Kind) {
+			continue
+		}
+
+		crds = append(crds, crd)
+	}
+
+	return crds, nil
+}
+
+// UpgradeCRDs upgrades custom resource definitions a directory containing CRD
+// YAML manifest files. It only supports
+// apiextensionsv1.CustomResourceDefinition. If the CRD exists in the cluster, it
+// will be patched from the contents from the corresponding manifest file. If the
+// CRD does not exist, it will be created. The Client must have the
+// apiextensionsv1.Scheme registered.
+func UpgradeCRDs(ctx context.Context, c ctrlclient.Client, dir string) error {
+	logger := zap.New().WithName("UpgradeCRDs").WithValues(
+		"version", version.Version(), "dir", dir)
+
+	crds, err := LoadCRDsFromDir(dir)
+	if err != nil {
+		return err
+	}
+
+	if len(crds) == 0 {
+		return fmt.Errorf("no CRDs found in directory %q", dir)
+	}
+
+	// TODO(future): add support for optionally deleting obsolete CRDs
+	var errs error
+	for _, crd := range crds {
+		var cur apiextensionsv1.CustomResourceDefinition
+		if err := c.Get(ctx, ctrlclient.ObjectKey{Name: crd.Name}, &cur); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Info("Creating", "name", crd.Name, "gvk", crd.GroupVersionKind())
+				if err := c.Create(ctx, &crd); err != nil {
+					return err
+				}
+			} else {
+				errs = errors.Join(errs, err)
+			}
+		} else {
+			logger.Info("Patching",
+				"name", crd.Name, "gvk", crd.GroupVersionKind(),
+				"uid", cur.GetUID(), "resourceVersion", cur.GetResourceVersion(),
+			)
+			patch := ctrlclient.MergeFrom(cur.DeepCopy())
+			cur.Spec = crd.Spec
+			cur.ObjectMeta.Annotations = crd.ObjectMeta.Annotations
+			cur.ObjectMeta.Labels = crd.ObjectMeta.Labels
```

Comment:
Any need to patch `cur.Status` as well?

Reply:
I don't think it should be necessary. The status should automatically be updated after we apply the patch.